src/map.h

```diff
@@ -4,69 +4,82 @@
  */
 
 /*
- * C Implementation for C++ std::map using red-black tree.
- *
- * Any data type can be stored in a map, just like std::map.
- * A map instance requires the specification of two file types:
- *   1. the key;
- *   2. what data type the tree node will store;
- *
- * It will also require a comparison function to sort the tree.
+ * This map implementation is heavily modified based on the rb.h header file of
+ * jemalloc. The original rb.h file from jemalloc has been extensively adapted
+ * and tailored for this map implementation. Credit and thanks to jemalloc for
+ * providing the initial implementation and inspiration.
  */
 
 #pragma once
 
 #include <stdbool.h>
 #include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#define RB_MAX_DEPTH (sizeof(void *) << 4)
+
+/* Node structure */
+#define rb_node(x_type)           \
+    struct {                      \
+        x_type *left, *right_red; \
+    }
+
+typedef struct map_node map_node_t;
+struct map_node {
+    void *key;
+    void *val;
+    rb_node(map_node_t) link;
+};
+
+typedef struct rb_tree {
+    map_node_t *root;
+    int (*cmp)(const map_node_t *, const map_node_t *);
+    size_t key_size;
```

Comment:
I am not sure if `key_size` and `val_size` make sense because we do know the actual size of keys and values.